crates/matrix-sdk/src/sliding_sync.rs

```diff
@@ -0,0 +1,957 @@
+
+use std::{fmt::Debug, sync::Arc};
+
+use anyhow::{bail, Context};
+use futures_core::stream::Stream;
+use matrix_sdk_base::deserialized_responses::SyncResponse;
+use ruma::{
+    api::client::sync::sync_events::v4,
+    assign,
+    events::{AnySyncTimelineEvent, RoomEventType},
+    serde::Raw,
+    OwnedRoomId, UInt,
+};
+use url::Url;
+
+use crate::{config::RequestConfig, Client, Result};
+
+/// Define the state the SlidingSync View is in
+///
+/// The lifetime of a SlidingSync usually starts at a `Preload`, getting a fast
+/// response for the first given number of Rooms, then switches into
+/// `CatchingUp` during which the view fetches the remaining rooms, usually in
+/// order, some times in batches. Once that is ready, it switches into `Live`.
+///
+/// If the client has been offline for a while, though, the SlidingSync might
+/// return back to `CatchingUp` at any point.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum SlidingSyncState {
+    /// Hasn't started yet
+    Cold,
+    /// We are quickly preloading a preview of the most important rooms
+    Preload,
+    /// We are trying to load all remaining rooms, might be in batches
+    CatchingUp,
+    /// We are all caught up and now only sync the live responses.
+    Live,
+}
+
+impl Default for SlidingSyncState {
+    fn default() -> Self {
+        SlidingSyncState::Cold
+    }
+}
+
+#[allow(dead_code)]
+/// Define the mode by which the the SlidingSyncView is in fetching the data
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum SlidingSyncMode {
+    /// FullSync all rooms in the background, configured with batch size
+    FullSync,
+    /// Only sync the specific windows defined
+    Selective,
+}
+
+impl Default for SlidingSyncMode {
+    fn default() -> Self {
+        SlidingSyncMode::FullSync
+    }
+}
+
+/// The Entry in the sliding sync room list per sliding sync view
+#[derive(Clone, Debug)]
+pub enum RoomListEntry {
+    /// This entry isn't known at this point and thus considered `Empty`
+    Empty,
+    /// There was `OwnedRoomId` but since the server told us to invalid this
+    /// entry. it is considered stale
+    Invalidated(OwnedRoomId),
+    /// This Entry is followed with `OwnedRoomId`
+    Filled(OwnedRoomId),
+}
+
+impl RoomListEntry {
+    /// Is this entry empty or invalidated?
+    pub fn empty_or_invalidated(&self) -> bool {
+        matches! {&self, RoomListEntry::Empty | RoomListEntry::Invalidated(_)}
+    }
+
+    /// The inner room_id if given
+    pub fn room_id(&self) -> Option<OwnedRoomId> {
+        match &self {
+            RoomListEntry::Empty => None,
+            RoomListEntry::Invalidated(b) | RoomListEntry::Filled(b) => Some(b.clone()),
+        }
+    }
+
+    /// Ref to the inner room id, if given
+    pub fn as_ref(&self) -> Option<&OwnedRoomId> {
+        match &self {
+            RoomListEntry::Empty => None,
+            RoomListEntry::Invalidated(b) | RoomListEntry::Filled(b) => Some(b),
+        }
+    }
```

Comment:
I think this should just be one method `as_room_id` that returns `&RoomId`, and to get an owned one you can call `.map(ToOwned::to_owned)` afterwards.

Reply:
:+1: